Sql query with Union causes the database to fail with panic.

Version: image from docker cockroachdb/cockroach:v1.1.4.

Whats happened: 
I want to insert a new row into the table with one query and return the identifier. In the event of a conflict, the identifier of the existing row must be returned without making any changes. 
Therefore, I use the following query
`SELECT user_id FROM[INSERT INTO user_table(user_id) VALUES(1) ON CONFLICT(user_id) DO NOTHING RETURNING user_id] UNION SELECT user_id FROM user_table WHERE user_id=1`.
But instead I get an error and the database is failed with panic.

*  Execute.

```CREATE DATABASE test_database;
CREATE DATABASE test_database;
SET database = test_database;
CREATE TABLE user_table(row_id SERIAL PRIMARY KEY, user_name STRING UNIQUE);
SELECT row_id FROM[INSERT INTO user_table(user_name) VALUES('test_name') ON CONFLICT(user_name) DO NOTHING RETURNING row_id] UNION SELECT row_id FROM user_table WHERE user_name='test_name';
```

*  Errors in sql console.

```
driver: bad connection
warning: connection lost!
opening new connection: all session settings will be lost
```

*  Logs from the standard out.

```
* WARNING: RUNNING IN INSECURE MODE!
* 
* - Your cluster is open for any client that can access <all your IP addresses>.
* - Any user, even root, can log in without providing a password.
* - Any user, connecting as root, can read or write any data in your cluster.
* - There is no network encryption nor authentication, and thus no confidentiality.
* 
* Check out how to secure your cluster: https://www.cockroachlabs.com/docs/stable/secure-a-cluster.html
*
CockroachDB node starting at 2018-02-01 16:36:27.495145241 +0000 UTC (took 0.3s)
build:      CCL v1.1.4 @ 2018/01/08 17:32:42 (go1.8.3)
admin:      http://a1e5ea3ba242:8080
sql:        postgresql://root@a1e5ea3ba242:80?application_name=cockroach&sslmode=disable
logs:       /cockroach/cockroach-data/logs
store[0]:   path=/cockroach/cockroach-data
status:     initialized new cluster
clusterID:  e62c219d-a813-4904-9efa-71b8b1e3e923
nodeID:     1
*
* ERROR: [client=127.0.0.1:53510,user=root,n1] a SQL panic has occurred!
*
*
* ERROR: [n1] a panic has occurred!
*
panic while executing "SELECT row_id FROM[INSERT INTO user_table(user_name) VALUES('test_name') ON CONFLICT(user_name) DO NOTHING RETURNING row_id] UNION SELECT row_id FROM user_table WHERE user_name='test_name';": row index 0 out of range
goroutine 771 [running]:
runtime/debug.Stack(0x7f4e29bf48b0, 0xc42021a2a0, 0x3)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x79
github.com/cockroachdb/cockroach/pkg/util/log.ReportPanic(0x7f4e29bf48b0, 0xc42021a2a0, 0xc4200fc000, 0x1cc0c40, 0xc4203d8360, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:123 +0xda
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4204b6120, 0x7f4e29bf48b0, 0xc42021a2a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:179 +0xd4
panic(0x1cc0c40, 0xc4203d8360)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc4207d0600, 0x7f4e29bf48b0, 0xc420d340c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:423 +0xa1
panic(0x1cc0c40, 0xc4203d8360)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql.(*Session).maybeRecover(0xc4206d8000, 0x1eab097, 0x9, 0xc4206c616a, 0xbd)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/session.go:1381 +0x28d
panic(0x1be47a0, 0xc4203e3580)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowContainer).At(0xc420bcb980, 0x0, 0xc4213d43b0, 0xe467d9, 0xc4206d8140)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/row_container.go:275 +0x205
github.com/cockroachdb/cockroach/pkg/sql.(*insertNode).Values(0xc4202e6c60, 0x7f4e29aeb190, 0xc420fb9600, 0xc4203e2e50)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/insert.go:642 +0x4a
github.com/cockroachdb/cockroach/pkg/sql.(*unionNode).Values(0xc420b11a90, 0x7f4e29aeb190, 0xc420fb9600, 0xc4206d8208)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/union.go:141 +0x49
github.com/cockroachdb/cockroach/pkg/sql.forEachRow(0x7f4e29aeb190, 0xc420fb9600, 0xc4206d8208, 0x2d82780, 0xc420b11a90, 0xc4213d4490, 0x0, 0x159bc4f)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:2043 +0x93
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execClassic(0xc4204ba480, 0xc4206d8208, 0x2d82780, 0xc420b11a90, 0x2d8c4a0, 0xc4207d0600, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:2020 +0x2df
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmt(0xc4204ba480, 0x2d83b80, 0xc420ca4750, 0x0, 0x0, 0x0, 0x150f40fb37facda1, 0x1, 0xc420d1cf60, 0xc4206d8208, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:2167 +0x637
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn(0xc4204ba480, 0xc4206d8000, 0x2d83b80, 0xc420ca4750, 0x0, 0x0, 0x0, 0x150f40fb37facda1, 0x1, 0xc420d1cf60, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1806 +0x739
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execSingleStatement(0xc4204ba480, 0xc4206d8000, 0x2d83b80, 0xc420ca4750, 0x0, 0x0, 0x0, 0x150f40fb37facda1, 0x1, 0xc420d1cf60, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1330 +0x92c
github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt(0xc4204ba480, 0xc4206d8000, 0xc420fb9580, 0x1, 0x1, 0x0, 0x1, 0x1, 0x0, 0x2d85ea0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1200 +0x38d
github.com/cockroachdb/cockroach/pkg/sql.runWithAutoRetry(0xc4204ba480, 0xc4206d8000, 0xc420fb9580, 0x1, 0x1, 0x9f70101, 0x0, 0x0, 0x3ff0000000000000, 0x322c4a0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1023 +0x1fe
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execParsed(0xc4204ba480, 0xc4206d8000, 0xc420fb9580, 0x1, 0x1, 0x0, 0x0, 0xc420017300, 0x604df0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:840 +0x26f
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc4204ba480, 0xc4206d8000, 0xc4206c616a, 0xbd, 0x0, 0x0, 0xbd, 0xe96)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:761 +0x493
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc4204ba480, 0xc4206d8000, 0xc4206c616a, 0xbd, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:642 +0x129
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc4207d0600, 0xc4207d0628, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:612 +0x148
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc4207d0600, 0x7f4e29bf48b0, 0xc420d340c0, 0xc4203c0d70, 0x5400, 0xc42043a368, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:503 +0xbd0
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc42043a200, 0x7f4e29aeb190, 0xc42104a580, 0x2d8c680, 0xc420fd6000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:421 +0x96c
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func9.1(0x2d8c680, 0xc420fd6000)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:750 +0x149
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4204b6120, 0x7f4e29bf48b0, 0xc42021a2a0, 0xc420500000, 0x2d8c680, 0xc420fd6000, 0xc420e1d860)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:140 +0xa9
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
	/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:142 +0x267
*
* ERROR: [n1] Reported as error 25a0ccdc3b4e4c9fbce3b1760838cf0f
*
```
 
*  Logs from file.

```
root@a1e5ea3ba242:/cockroach# cat /cockroach/cockroach-data/logs/cockroach.a1e5ea3ba242.root.2018-02-01T16_36_27Z.000006.log
I180201 16:36:27.220814 1 util/log/clog.go:1031  [config] file created at: 2018/02/01 16:36:27
I180201 16:36:27.220814 1 util/log/clog.go:1031  [config] running on machine: a1e5ea3ba242
I180201 16:36:27.220814 1 util/log/clog.go:1031  [config] binary: CockroachDB CCL v1.1.4 (linux amd64, built 2018/01/08 17:32:42, go1.8.3)
I180201 16:36:27.220814 1 util/log/clog.go:1031  [config] arguments: [cockroach start --insecure --port 80]
I180201 16:36:27.220814 1 util/log/clog.go:1031  line format: [IWEF]yymmdd hh:mm:ss.uuuuuu goid file:line msg utf8=✓
W180201 16:36:27.220813 1 cli/start.go:777  RUNNING IN INSECURE MODE!
- Your cluster is open for any client that can access <all your IP addresses>.
- Any user, even root, can log in without providing a password.
- Any user, connecting as root, can read or write any data in your cluster.
- There is no network encryption nor authentication, and thus no confidentiality.
Check out how to secure your cluster: https://www.cockroachlabs.com/docs/stable/secure-a-cluster.html
I180201 16:36:27.221182 1 server/config.go:323  available memory from cgroups (8.0 EiB) exceeds system memory 7.3 GiB, using system memory
W180201 16:36:27.221446 1 cli/start.go:697  Using the default setting for --cache (128 MiB).
  A significantly larger value is usually needed for good performance.
  If you have a dedicated server a reasonable setting is --cache=25% (1.8 GiB).
I180201 16:36:27.221479 1 cli/start.go:785  CockroachDB CCL v1.1.4 (linux amd64, built 2018/01/08 17:32:42, go1.8.3)
I180201 16:36:27.321806 1 server/config.go:323  available memory from cgroups (8.0 EiB) exceeds system memory 7.3 GiB, using system memory
I180201 16:36:27.321823 1 server/config.go:436  system total memory: 7.3 GiB
I180201 16:36:27.321869 1 server/config.go:438  server configuration:
max offset                500000000
cache size                128 MiB
SQL memory pool size      128 MiB
scan interval             10m0s
scan max idle time        200ms
metrics sample interval   10s
event log enabled         true
linearizable              false
I180201 16:36:27.321918 50 cli/start.go:503  starting cockroach node
I180201 16:36:27.322349 50 storage/engine/rocksdb.go:411  opening rocksdb instance at "/cockroach/cockroach-data/local"
I180201 16:36:27.334036 50 storage/engine/rocksdb.go:411  opening rocksdb instance at "/cockroach/cockroach-data"
I180201 16:36:27.343377 50 server/config.go:542  [n?] 1 storage engine initialized
I180201 16:36:27.343389 50 server/config.go:544  [n?] RocksDB cache size: 128 MiB
I180201 16:36:27.343397 50 server/config.go:544  [n?] store 0: RocksDB, max size 0 B, max open file limit 1043576
I180201 16:36:27.351239 50 server/node.go:344  [n?] **** cluster e62c219d-a813-4904-9efa-71b8b1e3e923 has been created
I180201 16:36:27.351284 50 server/server.go:837  [n?] **** add additional nodes by specifying --join=a1e5ea3ba242:80
I180201 16:36:27.353115 50 storage/store.go:1204  [n1,s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180201 16:36:27.353522 50 server/node.go:461  [n1] initialized store [n1,s1]: disk (capacity=226 GiB, available=113 GiB, used=16 KiB, logicalBytes=3.2 KiB), ranges=1, leases=0, writes=0.00, bytesPerReplica={p10=3266.00 p25=3266.00 p50=3266.00 p75=3266.00 p90=3266.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00}
I180201 16:36:27.353568 50 server/node.go:326  [n1] node ID 1 initialized
I180201 16:36:27.353732 50 gossip/gossip.go:327  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"a1e5ea3ba242:80" > attrs:<> locality:<> ServerVersion:<major_val:1 minor_val:1 patch:0 unstable:0 > 
I180201 16:36:27.354007 50 storage/stores.go:303  [n1] read 0 node addresses from persistent storage
I180201 16:36:27.354236 50 server/node.go:606  [n1] connecting to gossip network to verify cluster ID...
I180201 16:36:27.354268 50 server/node.go:631  [n1] node connected via gossip and verified as part of cluster "e62c219d-a813-4904-9efa-71b8b1e3e923"
I180201 16:36:27.354330 50 server/node.go:403  [n1] node=1: started with [<no-attributes>=/cockroach/cockroach-data] engine(s) and attributes []
I180201 16:36:27.354530 50 sql/executor.go:408  [n1] creating distSQLPlanner with address {tcp a1e5ea3ba242:80}
I180201 16:36:27.358547 112 storage/replica_command.go:2734  [split,n1,s1,r1/1:/M{in-ax}] initiating a split of this range at key /System/"" [r2]
I180201 16:36:27.368609 50 server/server.go:948  [n1] starting http server at a1e5ea3ba242:8080
I180201 16:36:27.368632 50 server/server.go:949  [n1] starting grpc/postgres server at a1e5ea3ba242:80
I180201 16:36:27.368650 50 server/server.go:950  [n1] advertising CockroachDB node at a1e5ea3ba242:80
W180201 16:36:27.368685 50 sql/jobs/registry.go:156  [n1] unable to get node liveness: node not in the liveness table
I180201 16:36:27.382758 112 storage/replica_command.go:2734  [split,n1,s1,r2/1:/{System/-Max}] initiating a split of this range at key /System/NodeLiveness [r3]
I180201 16:36:27.393208 50 sql/event_log.go:102  [n1] Event: "alter_table", target: 12, info: {TableName:eventlog Statement:ALTER TABLE system.eventlog ALTER COLUMN "uniqueID" SET DEFAULT uuid_v4() User:node MutationID:0 CascadeDroppedViews:[]}
I180201 16:36:27.396208 112 storage/replica_command.go:2734  [split,n1,s1,r3/1:/{System/NodeL…-Max}] initiating a split of this range at key /System/NodeLivenessMax [r4]
I180201 16:36:27.403488 50 sql/lease.go:342  [n1] publish: descID=12 (eventlog) version=2 mtime=2018-02-01 16:36:27.403270789 +0000 UTC
I180201 16:36:27.409147 112 storage/replica_command.go:2734  [split,n1,s1,r4/1:/{System/NodeL…-Max}] initiating a split of this range at key /System/tsd [r5]
E180201 16:36:27.425083 44 storage/replica_proposal.go:538  [n1,s1,r4/1:/{System/NodeL…-Max}] could not load SystemConfig span: must retry later due to intent on SystemConfigSpan
I180201 16:36:27.425649 112 storage/replica_command.go:2734  [split,n1,s1,r5/1:/{System/tsd-Max}] initiating a split of this range at key /System/"tse" [r6]
I180201 16:36:27.442118 112 storage/replica_command.go:2734  [split,n1,s1,r6/1:/{System/tse-Max}] initiating a split of this range at key /Table/SystemConfigSpan/Start [r7]
I180201 16:36:27.450194 50 sql/event_log.go:102  [n1] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:node}
I180201 16:36:27.455621 112 storage/replica_command.go:2734  [split,n1,s1,r7/1:/{Table/System…-Max}] initiating a split of this range at key /Table/11 [r8]
I180201 16:36:27.468291 112 storage/replica_command.go:2734  [split,n1,s1,r8/1:/{Table/11-Max}] initiating a split of this range at key /Table/12 [r9]
I180201 16:36:27.479443 50 sql/event_log.go:102  [n1] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:1.1 User:node}
I180201 16:36:27.485616 112 storage/replica_command.go:2734  [split,n1,s1,r9/1:/{Table/12-Max}] initiating a split of this range at key /Table/13 [r10]
I180201 16:36:27.490731 50 sql/event_log.go:102  [n1] Event: "set_cluster_setting", target: 0, info: {SettingName:trace.debug.enable Value:false User:node}
I180201 16:36:27.495111 50 server/server.go:1092  [n1] done ensuring all necessary migrations have run
I180201 16:36:27.495126 50 server/server.go:1094  [n1] serving sql connections
I180201 16:36:27.495195 50 cli/start.go:582  node startup completed:
CockroachDB node starting at 2018-02-01 16:36:27.495145241 +0000 UTC (took 0.3s)
build:      CCL v1.1.4 @ 2018/01/08 17:32:42 (go1.8.3)
admin:      http://a1e5ea3ba242:8080
sql:        postgresql://root@a1e5ea3ba242:80?application_name=cockroach&sslmode=disable
logs:       /cockroach/cockroach-data/logs
store[0]:   path=/cockroach/cockroach-data
status:     initialized new cluster
clusterID:  e62c219d-a813-4904-9efa-71b8b1e3e923
nodeID:     1
I180201 16:36:27.499380 112 storage/replica_command.go:2734  [split,n1,s1,r10/1:/{Table/13-Max}] initiating a split of this range at key /Table/14 [r11]
I180201 16:36:27.505845 495 sql/event_log.go:102  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:a1e5ea3ba242:80} Attrs: Locality: ServerVersion:1.1} ClusterID:e62c219d-a813-4904-9efa-71b8b1e3e923 StartedAt:1517502987354272551 LastUp:1517502987354272551}
I180201 16:36:27.509991 112 storage/replica_command.go:2734  [split,n1,s1,r11/1:/{Table/14-Max}] initiating a split of this range at key /Table/15 [r12]
I180201 16:36:27.520885 112 storage/replica_command.go:2734  [split,n1,s1,r12/1:/{Table/15-Max}] initiating a split of this range at key /Table/16 [r13]
I180201 16:36:27.530588 112 storage/replica_command.go:2734  [split,n1,s1,r13/1:/{Table/16-Max}] initiating a split of this range at key /Table/17 [r14]
I180201 16:36:27.550841 112 storage/replica_command.go:2734  [split,n1,s1,r14/1:/{Table/17-Max}] initiating a split of this range at key /Table/18 [r15]
I180201 16:36:27.571611 112 storage/replica_command.go:2734  [split,n1,s1,r15/1:/{Table/18-Max}] initiating a split of this range at key /Table/19 [r16]
I180201 16:36:37.356021 60 server/status/runtime.go:223  [n1] runtime stats: 59 MiB RSS, 124 goroutines, 12 MiB/2.9 MiB/24 MiB GO alloc/idle/total, 12 MiB/16 MiB CGO alloc/total, 0.00cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (8x)
I180201 16:36:37.356296 57 storage/store.go:4183  [n1,s1] sstables (read amplification = 0):
I180201 16:36:37.356559 57 storage/store.go:4184  [n1,s1] 
** Compaction Stats [default] **
Level    Files   Size     Score Read(GB)  Rn(GB) Rnp1(GB) Write(GB) Wnew(GB) Moved(GB) W-Amp Rd(MB/s) Wr(MB/s) Comp(sec) Comp(cnt) Avg(sec) KeyIn KeyDrop
----------------------------------------------------------------------------------------------------------------------------------------------------------
 Sum      0/0    0.00 KB   0.0      0.0     0.0      0.0       0.0      0.0       0.0   0.0      0.0      0.0         0         0    0.000       0      0
 Int      0/0    0.00 KB   0.0      0.0     0.0      0.0       0.0      0.0       0.0   0.0      0.0      0.0         0         0    0.000       0      0
Uptime(secs): 10.0 total, 10.0 interval
Flush(GB): cumulative 0.000, interval 0.000
AddFile(GB): cumulative 0.000, interval 0.000
AddFile(Total Files): cumulative 0, interval 0
AddFile(L0 Files): cumulative 0, interval 0
AddFile(Keys): cumulative 0, interval 0
Cumulative compaction: 0.00 GB write, 0.00 MB/s write, 0.00 GB read, 0.00 MB/s read, 0.0 seconds
Interval compaction: 0.00 GB write, 0.00 MB/s write, 0.00 GB read, 0.00 MB/s read, 0.0 seconds
Stalls(count): 0 level0_slowdown, 0 level0_slowdown_with_compaction, 0 level0_numfiles, 0 level0_numfiles_with_compaction, 0 stop for pending_compaction_bytes, 0 slowdown for pending_compaction_bytes, 0 memtable_compaction, 0 memtable_slowdown, interval 0 total count
I180201 16:36:37.359730 117 storage/replica_proposal.go:453  [replicate,n1,s1,r10/1:/Table/1{3-4}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.357076711,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:37.362338 122 storage/replica_proposal.go:453  [replicate,n1,s1,r14/1:/Table/1{7-8}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.360253594,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:37.364793 118 storage/replica_proposal.go:453  [replicate,n1,s1,r15/1:/Table/1{8-9}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.362824006,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:37.367038 126 storage/replica_proposal.go:453  [replicate,n1,s1,r9/1:/Table/1{2-3}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.365122975,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:37.371349 124 storage/replica_proposal.go:453  [replicate,n1,s1,r12/1:/Table/1{5-6}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.369545324,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:37.373835 120 storage/replica_proposal.go:453  [replicate,n1,s1,r13/1:/Table/1{6-7}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.371727856,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:37.376314 127 storage/replica_proposal.go:453  [replicate,n1,s1,r16/1:/{Table/19-Max}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.374226482,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:37.379330 123 storage/replica_proposal.go:453  [replicate,n1,s1,r6/1:/{System/tse-Table/System…}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.376665352,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:37.383940 133 storage/replica_proposal.go:453  [n1,s1,r4/1:/System/{NodeLive…-tsd}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.381827301,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:37.385242 49 storage/replica_proposal.go:453  [replicate,n1,s1,r8/1:/Table/1{1-2}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.382108770,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:37.388300 115 storage/replica_proposal.go:453  [replicate,n1,s1,r7/1:/Table/{SystemCon…-11}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.385926791,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:37.390971 132 storage/replica_proposal.go:453  [replicate,n1,s1,r11/1:/Table/1{4-5}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.388989455,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:37.393217 131 storage/replica_proposal.go:453  [replicate,n1,s1,r5/1:/System/ts{d-e}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.391314687,0 following repl=(n1,s1):1 start=0.000000000,0 exp=1517502996.352243591,0 pro=1517502987.352285108,0
I180201 16:36:47.355661 60 server/status/runtime.go:223  [n1] runtime stats: 62 MiB RSS, 124 goroutines, 18 MiB/480 KiB/27 MiB GO alloc/idle/total, 12 MiB/16 MiB CGO alloc/total, 106.20cgo/sec, 0.01/0.00 %(u/s)time, 0.00 %gc (0x)
I180201 16:36:57.355639 60 server/status/runtime.go:223  [n1] runtime stats: 64 MiB RSS, 124 goroutines, 14 MiB/5.3 MiB/29 MiB GO alloc/idle/total, 12 MiB/16 MiB CGO alloc/total, 57.20cgo/sec, 0.01/0.00 %(u/s)time, 0.00 %gc (1x)
I180201 16:37:07.355753 60 server/status/runtime.go:223  [n1] runtime stats: 65 MiB RSS, 125 goroutines, 12 MiB/7.6 MiB/29 MiB GO alloc/idle/total, 12 MiB/16 MiB CGO alloc/total, 58.80cgo/sec, 0.01/0.00 %(u/s)time, 0.00 %gc (1x)
I180201 16:37:14.801036 771 sql/event_log.go:102  [client=127.0.0.1:53510,user=root,n1] Event: "create_database", target: 50, info: {DatabaseName:test_database Statement:CREATE DATABASE test_database User:root}
I180201 16:37:14.803152 112 storage/replica_command.go:2734  [split,n1,s1,r16/1:/{Table/19-Max}] initiating a split of this range at key /Table/50 [r17]
I180201 16:37:14.822803 128 storage/replica_proposal.go:453  [n1,s1,r16/1:/{Table/19-Max}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.374226482,0 following repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.374226482,0
I180201 16:37:16.273532 771 sql/event_log.go:102  [client=127.0.0.1:53510,user=root,n1] Event: "create_table", target: 51, info: {TableName:user_table Statement:CREATE TABLE user_table (row_id SERIAL PRIMARY KEY, user_name STRING UNIQUE) User:root}
I180201 16:37:16.274399 112 storage/replica_command.go:2734  [split,n1,s1,r17/1:/{Table/50-Max}] initiating a split of this range at key /Table/51 [r18]
I180201 16:37:16.291939 115 storage/replica_proposal.go:453  [n1,s1,r17/1:/{Table/50-Max}] new range lease repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.374226482,0 following repl=(n1,s1):1 start=0.000000000,0 epo=1 pro=1517502997.374226482,0
I180201 16:37:17.355124 60 server/status/runtime.go:223  [n1] runtime stats: 66 MiB RSS, 125 goroutines, 20 MiB/1.3 MiB/29 MiB GO alloc/idle/total, 12 MiB/16 MiB CGO alloc/total, 161.31cgo/sec, 0.02/0.01 %(u/s)time, 0.00 %gc (0x)
E180201 16:37:27.181628 771 sql/session.go:1370  [client=127.0.0.1:53510,user=root,n1] a SQL panic has occurred!
E180201 16:37:27.181934 771 util/log/crash_reporting.go:113  [n1] a panic has occurred!
I180201 16:37:27.355104 58 gossip/gossip.go:478  [n1] gossip status (ok, 1 node)
gossip client (0/3 cur/max conns)
gossip server (0/3 cur/max conns, infos 0/0 sent/received, bytes 0B/0B sent/received)
I180201 16:37:27.356163 60 server/status/runtime.go:223  [n1] runtime stats: 66 MiB RSS, 130 goroutines, 16 MiB/4.9 MiB/30 MiB GO alloc/idle/total, 12 MiB/16 MiB CGO alloc/total, 76.09cgo/sec, 0.01/0.00 %(u/s)time, 0.00 %gc (1x)
E180201 16:37:28.108400 771 util/log/crash_reporting.go:366  [n1] Reported as error 25a0ccdc3b4e4c9fbce3b1760838cf0f
panic: row index 0 out of range [recovered]
	panic: panic while executing "SELECT row_id FROM[INSERT INTO user_table(user_name) VALUES('test_name') ON CONFLICT(user_name) DO NOTHING RETURNING row_id] UNION SELECT row_id FROM user_table WHERE user_name='test_name';": row index 0 out of range [recovered]
	panic: panic while executing "SELECT row_id FROM[INSERT INTO user_table(user_name) VALUES('test_name') ON CONFLICT(user_name) DO NOTHING RETURNING row_id] UNION SELECT row_id FROM user_table WHERE user_name='test_name';": row index 0 out of range [recovered]
	panic: panic while executing "SELECT row_id FROM[INSERT INTO user_table(user_name) VALUES('test_name') ON CONFLICT(user_name) DO NOTHING RETURNING row_id] UNION SELECT row_id FROM user_table WHERE user_name='test_name';": row index 0 out of range
goroutine 771 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4204b6120, 0x7f4e29bf48b0, 0xc42021a2a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:181 +0x98
panic(0x1cc0c40, 0xc4203d8360)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc4207d0600, 0x7f4e29bf48b0, 0xc420d340c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:423 +0xa1
panic(0x1cc0c40, 0xc4203d8360)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql.(*Session).maybeRecover(0xc4206d8000, 0x1eab097, 0x9, 0xc4206c616a, 0xbd)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/session.go:1381 +0x28d
panic(0x1be47a0, 0xc4203e3580)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowContainer).At(0xc420bcb980, 0x0, 0xc4213d43b0, 0xe467d9, 0xc4206d8140)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/row_container.go:275 +0x205
github.com/cockroachdb/cockroach/pkg/sql.(*insertNode).Values(0xc4202e6c60, 0x7f4e29aeb190, 0xc420fb9600, 0xc4203e2e50)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/insert.go:642 +0x4a
github.com/cockroachdb/cockroach/pkg/sql.(*unionNode).Values(0xc420b11a90, 0x7f4e29aeb190, 0xc420fb9600, 0xc4206d8208)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/union.go:141 +0x49
github.com/cockroachdb/cockroach/pkg/sql.forEachRow(0x7f4e29aeb190, 0xc420fb9600, 0xc4206d8208, 0x2d82780, 0xc420b11a90, 0xc420dca490, 0x0, 0x159bc4f)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:2043 +0x93
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execClassic(0xc4204ba480, 0xc4206d8208, 0x2d82780, 0xc420b11a90, 0x2d8c4a0, 0xc4207d0600, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:2020 +0x2df
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmt(0xc4204ba480, 0x2d83b80, 0xc420ca4750, 0x0, 0x0, 0x0, 0x150f40fb37facda1, 0x1, 0xc420d1cf60, 0xc4206d8208, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:2167 +0x637
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn(0xc4204ba480, 0xc4206d8000, 0x2d83b80, 0xc420ca4750, 0x0, 0x0, 0x0, 0x150f40fb37facda1, 0x1, 0xc420d1cf60, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1806 +0x739
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execSingleStatement(0xc4204ba480, 0xc4206d8000, 0x2d83b80, 0xc420ca4750, 0x0, 0x0, 0x0, 0x150f40fb37facda1, 0x1, 0xc420d1cf60, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1330 +0x92c
github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt(0xc4204ba480, 0xc4206d8000, 0xc420fb9580, 0x1, 0x1, 0x0, 0x1, 0x1, 0x0, 0x2d85ea0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1200 +0x38d
github.com/cockroachdb/cockroach/pkg/sql.runWithAutoRetry(0xc4204ba480, 0xc4206d8000, 0xc420fb9580, 0x1, 0x1, 0x9f70101, 0x0, 0x0, 0x3ff0000000000000, 0x322c4a0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1023 +0x1fe
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execParsed(0xc4204ba480, 0xc4206d8000, 0xc420fb9580, 0x1, 0x1, 0x0, 0x0, 0xc420017300, 0x604df0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:840 +0x26f
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc4204ba480, 0xc4206d8000, 0xc4206c616a, 0xbd, 0x0, 0x0, 0xbd, 0xe96)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:761 +0x493
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc4204ba480, 0xc4206d8000, 0xc4206c616a, 0xbd, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:642 +0x129
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc4207d0600, 0xc4207d0628, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:612 +0x148
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc4207d0600, 0x7f4e29bf48b0, 0xc420d340c0, 0xc4203c0d70, 0x5400, 0xc42043a368, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:503 +0xbd0
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc42043a200, 0x7f4e29aeb190, 0xc42104a580, 0x2d8c680, 0xc420fd6000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:421 +0x96c
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func9.1(0x2d8c680, 0xc420fd6000)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:750 +0x149
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4204b6120, 0x7f4e29bf48b0, 0xc42021a2a0, 0xc420500000, 0x2d8c680, 0xc420fd6000, 0xc420e1d860)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:140 +0xa9
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
	/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:142 +0x267
```

Useful information.
1. After restoring the database, the desired record is in the table. Thus, the query actually inserts a new row into the database.
2. Next attempts to insert this row are successful: returns the id of an existing row, no changes performed.

--- a/pkg/sql/insert.go
+++ b/pkg/sql/insert.go
@@ -365,7 +365,12 @@ func (n *insertNode) startExec(params runParams) error {
 
 func (n *insertNode) Next(params runParams) (bool, error) {
 	if n.run.isUpsertReturning {
-		return n.drain(params)
+		hasRows, err := n.drain(params)
+		if hasRows {
+			n.run.resultRow = n.run.rowsUpserted.At(0)
+			n.run.rowsUpserted.PopFirst()
+		}
+		return hasRows, err
 	}
 
 	return n.internalNext(params)
@@ -392,13 +397,7 @@ func (n *insertNode) Close(ctx context.Context) {
 }
 
 func (n *insertNode) Values() tree.Datums {
-	if !n.run.isUpsertReturning {
-		return n.run.resultRow
-	}
-
-	row := n.run.rowsUpserted.At(0)
-	n.run.rowsUpserted.PopFirst()
-	return row
+	return n.run.resultRow
 }
 
 // Because TableUpserter batches the upserts, we need to completely drain the

--- a/pkg/sql/logictest/testdata/logic_test/upsert
+++ b/pkg/sql/logictest/testdata/logic_test/upsert
@@ -228,6 +228,13 @@ upsert INTO upsert_returning VALUES (1, 5, 4, 3), (6, 5, 4, 3) RETURNING *
 statement ok
 COMMIT
 
+# For #22300. Test UPSERT ... RETURNING with UNION.
+query I rowsort
+SELECT a FROM [UPSERT INTO upsert_returning VALUES (7) RETURNING a] UNION VALUES (8)
+----
+7
+8
+
 # For #6710. Add an unused column to disable the fast path which doesn't have this bug.
 statement ok
 CREATE TABLE issue_6710 (a INT PRIMARY KEY, b STRING, c INT)

